src/Agent.Worker/Build/TfsVCTeeSourceProvider.cs

```diff
@@ -0,0 +1,373 @@
+using Microsoft.TeamFoundation.Build.WebApi;
+using Microsoft.TeamFoundation.DistributedTask.WebApi;
+using Microsoft.VisualStudio.Services.Agent;
+using Microsoft.VisualStudio.Services.Agent.Util;
+using Microsoft.VisualStudio.Services.Agent.Worker;
+using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker.Build
+{
+    public sealed class TfsVCTeeSourceProvider : SourceProvider, ISourceProvider
+    {
+        private string _workspaceName;
```

Comment:
I don't like the mess I've made here in this class. In a follow-up PR (to make shelvesets work), I'll remove the dependency this class has on the tracking config and use variables dictionary instead. I think the workspace name needs to go into the variable dictionary anyway, if so I'll move it as part of that PR. I need to dump the variables from a build on an old Windows agent and make sure I'm not missing setting any variables.